packages/connect/src/protocol-connect/handler-factory.spec.ts

```diff
@@ -51,51 +53,89 @@ describe("createHandlerFactory()", function () {
-        implDefault = async function (req: Message, ctx: HandlerContext) {
-          ctx.responseHeader.set("stub-handler", "1");
-          return new ctx.method.O();
-        } as unknown as MethodImpl<M>;
-        break;
-      case MethodKind.ServerStreaming:
-        // eslint-disable-next-line @typescript-eslint/require-await
-        implDefault = async function* (req: Message, ctx: HandlerContext) {
-          ctx.responseHeader.set("stub-handler", "1");
-          yield new ctx.method.O();
-        } as unknown as MethodImpl<M>;
-        break;
-      case MethodKind.ClientStreaming:
-      case MethodKind.BiDiStreaming:
-        implDefault = (() => {
-          throw new Error("not implemented");
-        }) as unknown as MethodImpl<M>;
-        break;
-    }
-    const spec = createMethodImplSpec(
-      opt.service ?? testService,
-      method,
-      opt.impl ?? implDefault
+    const h = createHandlerFactory(opt)(
+      createMethodImplSpec(testService, method, impl)
     );
-    const f = createHandlerFactory(opt);
-    return f(spec);
+    const t = createTransport({
+      httpClient: createUniversalHandlerClient([h]),
+      baseUrl: "https://example.com",
+      readMaxBytes: 0xffffff,
+      writeMaxBytes: 0xffffff,
+      compressMinBytes: 0xffffff,
+      useBinaryFormat: true,
+      interceptors: [],
+      acceptCompression: [],
+      sendCompression: null,
+    });
+    return {
+      service: testService,
+      method: method,
+      handler: h,
+      transport: t,
+    };
   }
 
+  describe("returned handler", function () {
+    it("should surface headers for unary", async function () {
+      const { transport, service, method } = setupTestHandler(
+        testService.methods.foo,
+        {},
+        (req, ctx) => {
+          ctx.responseHeader.set("implementation-called", "yes");
+          return { value: req.value.toString(10) };
+        }
+      );
+      const r = await transport.unary(
+        service,
+        method,
+        undefined,
+        undefined,
+        undefined,
+        new Int32Value({ value: 123 })
+      );
+      expect(r.header.get("implementation-called")).toBe("yes");
+      expect(r.message.value).toBe("123");
+    });
+
+    it("should surface headers for server-streaming", async function () {
+      const { transport, service, method } = setupTestHandler(
+        testService.methods.bar,
+        {},
+        // eslint-disable-next-line @typescript-eslint/require-await
+        async function* (req, ctx) {
+          ctx.responseHeader.set("implementation-called", "yes");
+          yield { value: req.value.toString(10) };
+        }
+      );
+      const r = await transport.stream(
+        service,
+        method,
+        undefined,
+        undefined,
+        undefined,
+        createAsyncIterable([new Int32Value({ value: 123 })])
+      );
+      expect(r.header.get("implementation-called")).toBe("yes");
```

Comment:
We are testing that response headers set by the implementation are present in the response received by the client, in all three protocols, for unary and for server-streaming.

Reply:
looks great.  I like that we're also checking that the header value is correct at the same time.  The current approach used in `interceptor.spec.ts` should probably be improved to do the same thing because right now the "`log` vs `wantLog`" implementation just checks header presence without inspecting the values.

Reply:
Wouldn't hurt, yes.